[rpc][easy] move unnecessary python call directly to pybind

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#33174 [rpc][easy] move unnecessary python call directly to pybind**

Closes https://github.com/pytorch/pytorch/issues/32780. It looks like
this is the only callsite where we do `_get_current_rpc_agent().foo()`, and we
can do this directly in the pybind layer to save some overhead.

Differential Revision: [D19828786](https://our.internmc.facebook.com/intern/diff/D19828786/)

rohan-varma committed:
[rpc][easy] move unnecessary python call directly to pybind

Closes https://github.com/pytorch/pytorch/issues/32780. It looks like
this is the only callsite where we do `_get_current_rpc_agent().foo()`, and we
can do this directly in the pybind layer to save some overhead.

Differential Revision: [D19828786](https://our.internmc.facebook.com/intern/diff/D19828786/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc][easy] move unnecessary python call directly to pybind"


Closes https://github.com/pytorch/pytorch/issues/32780. It looks like
this is the only callsite where we do `_get_current_rpc_agent().foo()`, and we
can do this directly in the pybind layer to save some overhead.

Differential Revision: [D19828786](https://our.internmc.facebook.com/intern/diff/D19828786/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -414,6 +410,19 @@ If the future completes with an error, an exception is thrown.
+
+    Arguments:
+        flag (bool): True to set GIL profiling, False to disable.
+      )");
```
the old `enable_gil_profiling` used to release GIL, but no longer so in the new API. I guess the new version is the right approach, as it would be better to holding GIL for super light method to avoid ctx switching cost.

mrshenli reviewed (approved):
LGTM. Test failures look irrelevant.

rohan-varma committed:
Update on "[rpc][easy] move unnecessary python call directly to pybind"


Closes https://github.com/pytorch/pytorch/issues/32780. It looks like
this is the only callsite where we do `_get_current_rpc_agent().foo()`, and we
can do this directly in the pybind layer to save some overhead.

Differential Revision: [D19828786](https://our.internmc.facebook.com/intern/diff/D19828786/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc][easy] move unnecessary python call directly to pybind"


Closes https://github.com/pytorch/pytorch/issues/32780. It looks like
this is the only callsite where we do `_get_current_rpc_agent().foo()`, and we
can do this directly in the pybind layer to save some overhead.

Differential Revision: [D19828786](https://our.internmc.facebook.com/intern/diff/D19828786/)

[ghstack-poisoned]
